troubleshooting/checkpointing/src/solution/java/com/ververica/flink/training/solutions/CheckpointingJobSolution3.java

```diff
@@ -0,0 +1,297 @@
+
+/**
+ * Solution 3 fixes the streaming job with slow checkpointing by sorting the stream based on event time
+ * then pre-aggregation.
+ *
+ * In addition, the event time timestamp is accessible in the sorting operator and not needed in the window operator
+ * anymore, this solution also change to use DataStream<Measurement> instead of DataStream<Tuple2<Measurement, Long>>
+ */
+public class CheckpointingJobSolution3 {
+
+	/**
+	 * Creates and starts the troubled streaming job.
+	 *
+	 * @throws Exception if the application is misconfigured or fails during job submission
+	 */
+	public static void main(String[] args) throws Exception {
+		ParameterTool parameters = ParameterTool.fromArgs(args);
+
+		StreamExecutionEnvironment env = createConfiguredEnvironment(parameters);
+
+		//Timing Configuration
+		env.getConfig().setAutoWatermarkInterval(100);
+		env.setBufferTimeout(10);
+
+		//Checkpointing Configuration (use cluster-configs if not run locally)
+		if (isLocal(parameters)) {
+			env.enableCheckpointing(TimeUnit.SECONDS.toMillis(10));
+			env.getCheckpointConfig().setMinPauseBetweenCheckpoints(TimeUnit.SECONDS.toMillis(10));
+			env.getCheckpointConfig().setCheckpointTimeout(TimeUnit.MINUTES.toMillis(2));
+		}
+
+		DataStream<Measurement> sourceStream = env
+				.addSource(SourceUtils.createFailureFreeFakeKafkaSource())
+				.name("FakeKafkaSource")
+				.uid("FakeKafkaSource")
+				.assignTimestampsAndWatermarks(
+						WatermarkStrategy
+								.<FakeKafkaRecord>forBoundedOutOfOrderness(Duration.ofMillis(250))
+								.withTimestampAssigner(
+										(element, timestamp) -> element.getTimestamp())
+								.withIdleness(Duration.ofSeconds(1)))
+				.name("Watermarks")
+				.uid("Watermarks")
+				.flatMap(new MeasurementDeserializer())
+				.name("Deserialization")
+				.uid("Deserialization");
+
+		DataStream<Measurement> sortedStream = sourceStream
+				.keyBy(Measurement::getSensorId)
+				.process(new SortMeasurementFunction())
+				.name("Sorting")
+				.uid("Sorting");
+
+		KeyedStream<Measurement, Integer> keyedSortedStream =
+				DataStreamUtils.reinterpretAsKeyedStream(
+						sortedStream,
+						Measurement::getSensorId);
+
+		DataStream<WindowedMeasurements> aggregatedPerLocation = keyedSortedStream
+				.window(SlidingEventTimeWindows.of(Time.of(1, TimeUnit.MINUTES), Time.of(1, TimeUnit.SECONDS)))
+				.aggregate(new MeasurementWindowAggregatingFunction(),
+						new MeasurementWindowProcessFunction())
+				.name("WindowedAggregationPerLocation")
+				.uid("WindowedAggregationPerLocation");
+
+		if (isLocal(parameters)) {
+			aggregatedPerLocation.print()
+					.name("NormalOutput")
+					.uid("NormalOutput")
+					.disableChaining();
+		} else {
+			aggregatedPerLocation.addSink(new DiscardingSink<>())
+					.name("NormalOutput")
+					.uid("NormalOutput")
+					.disableChaining();
+		}
+
+		env.execute(CheckpointingJobSolution3.class.getSimpleName());
+	}
+
+	public static class SortMeasurementFunction
```

Comment:
Can you add a javadoc for this class that describes on a high level what it is doing and how?